File: demos/demo_LbBruteforce_DTW.py
import sys
import os
import numpy as np

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from daisy import DistanceType, LbBruteforce

def main():
    n_database = 200000
    dim = 96
    n_query = 10
    k = 5

    # 1. Generate random data and queries
    np.random.seed(100)   
    db = np.random.randn(n_database, dim).astype(np.float32)

    np.random.seed(50)
    query = np.random.randn(n_query, dim).astype(np.float32)

    # 2. Create a LbBruteforce search object with DTW distance
    index = LbBruteforce(DistanceType.DTW)
    print("Created LbBruteforce with DTW distance type")

    # 3. Build the index
    index.setNumThreads(1)
    index.buildIndex(db)
    print("Index built successfully")

    # 4. Search the index
    print("Starting DTW search...")
    I, D = index.searchIndex(query, k)
    print("DTW search completed successfully!")

    for query_num in range(n_query):
        print(f"Query {query_num}:")
        print("Distances:", D[query_num])
        print("Indices:", I[query_num])
        print()

    print("Python DTW demo completed successfully!")

if __name__ == "__main__":
    main()

File: demos/demo_LbBruteforce_DTW.cpp
#include "../commons/dataloaders.hpp"
#include "../lib/algos/LbBruteforce.hpp"

int main()
{
    // 0. Configuration of the variables
    diNoLib::idx_t n_database = 200000;
    unsigned long long dim = 96;
    unsigned long long n_query = 10;
    diNoLib::idx_t k = 5;

    // 1. Generate random data and queries
    float *database = loadRandomData(n_database, dim, 100, true);
    float *query = loadRandomData(n_query, dim, 50, true);

    printf("Loaded %llu database points and %llu query points with dimension %llu\n", n_database, n_query, dim);

    // 2. Create a LbBruteforce search object
    diNoLib::LbBruteforce bf_search(diNoLib::DistanceType::DTW);

    // 3. Build the index (simplified API - no need for DataSource!)
    bf_search.buildIndex(database, n_database, dim);

    // 4. Search the index
    diNoLib::idx_t *I = new diNoLib::idx_t[n_query * k];
    float *D = new float[n_query * k];
    bf_search.searchIndex(query, n_query, k, I, D);

    // 5. Print the results
    for (diNoLib::idx_t i = 0; i < n_query; i++)
    {
        printf("Query %llu: ", i);
        for (diNoLib::idx_t j = 0; j < k; j++)
        {
            printf("%llu ", I[i * k + j]);
        }
        printf("\n");
    }

    // 6. Clean up
    delete[] database;
    delete[] query;
    delete[] I;
    delete[] D;

    return 0;
}